classes/village/VillageUpgradeManager.php

```diff
@@ -0,0 +1,368 @@
+            $building->construction_time_remaining = System::TimeRemaining($building->construction_progress_required - $building->construction_progress, format: "long", include_seconds: false, include_minutes: false);
+            $building->requirements_met = VillageUpgradeManager::checkConstructionRequirementsMet($village, $building->key, $building->tier + 1);
+            foreach (VillageBuildingConfig::BUILDING_UPGRADE_SETS[$building->key] as $upgrade_set_key) {
+                $upgrade_set = new VillageUpgradeSetDto(
+                    key: $upgrade_set_key,
+                    name: VillageBuildingConfig::UPGRADE_SET_NAMES[$upgrade_set_key],
+                    description: VillageBuildingConfig::UPGRADE_SET_DESCRIPTIONS[$upgrade_set_key],
+                );
+                foreach (VillageBuildingConfig::UPGRADE_SET_UPGRADES[$upgrade_set_key] as $upgrade_key) {
+                    $upgrade = new VillageUpgradeDto(
+                        id: $village->upgrades[$upgrade_key]->id ?? null,
+                        key: $upgrade_key,
+                        village_id: $village->village_id,
+                        status: $village->upgrades[$upgrade_key]->status ?? VillageUpgradeConfig::UPGRADE_STATUS_LOCKED,
+                        research_progress: $village->upgrades[$upgrade_key]->research_progress ?? null,
+                        research_progress_required: $village->upgrades[$upgrade_key]->research_progress_required ?? null,
+                        name: VillageUpgradeConfig::UPGRADE_NAMES[$upgrade_key],
+                        description: VillageUpgradeConfig::UPGRADE_DESCRIPTIONS[$upgrade_key],
+                        materials_research_cost: VillageUpgradeConfig::UPGRADE_RESEARCH_COST[$upgrade_key][WarManager::RESOURCE_MATERIALS],
+                        food_research_cost: VillageUpgradeConfig::UPGRADE_RESEARCH_COST[$upgrade_key][WarManager::RESOURCE_FOOD],
+                        wealth_research_cost: VillageUpgradeConfig::UPGRADE_RESEARCH_COST[$upgrade_key][WarManager::RESOURCE_WEALTH],
+                        research_time: VillageUpgradeConfig::UPGRADE_RESEARCH_TIME[$upgrade_key],
+                        research_time_remaining: isset($village->upgrades[$upgrade_key]->research_progress_last_updated) ? System::TimeRemaining($village->upgrades[$upgrade_key]->research_progress_required - $village->upgrades[$upgrade_key]->research_progress, format: "long", include_seconds: false, include_minutes: false) : null,
+                        food_upkeep: VillageUpgradeConfig::UPGRADE_UPKEEP[$upgrade_key][WarManager::RESOURCE_FOOD],
+                        materials_upkeep: VillageUpgradeConfig::UPGRADE_UPKEEP[$upgrade_key][WarManager::RESOURCE_MATERIALS],
+                        wealth_upkeep: VillageUpgradeConfig::UPGRADE_UPKEEP[$upgrade_key][WarManager::RESOURCE_WEALTH],
+                        requirements_met: VillageUpgradeManager::checkResearchRequirementsMet($village, $upgrade_key),
```

Comment:
Had an idea here:
- Split `VillageUpgradeDto` into two classes:
  - `VillageUpgrade` represents the config + the current status for this village. This is the main class we work with in application code. The idea is it's never missing data, if you have an instance of `VillageUpgrade`, you have everything you need
  - `VillageUpgradeDb` / `VillageUpgradeData` represents just the status for this village loaded from the db in `getUpgradesForVillage`
- Move this initialize logic to a helper in `VillageUpgrade`

```php

public static function getBuildingUpgradesForDisplay(System $system, Village $village): array {
  // ..
  foreach (VillageBuildingConfig::UPGRADE_SET_UPGRADES[$upgrade_set_key] as $upgrade_key) {
    $upgrade = VillageUpgrade::fromUpgradeData(
      key: $upgrade_key, 
      village_id: $village->id, 
      upgrade_data: $village->upgrade_data[$upgrade_key]
    );
  }
  // ..
}

// ...

class VillageUpgrade {
  // ...
  public static function init(string $upgrade_key, int $village_id, VillageUpgradeData $upgrade_data): VillageUpgrade {
      return new VillageUpgradeDto(
      id: $upgrade_data->id ?? null,
      key: $upgrade_key,
      village_id: $village_id,
      status: $upgrade_data->status ?? VillageUpgradeConfig::UPGRADE_STATUS_LOCKED,
      // ..
      name: VillageUpgradeConfig::UPGRADE_NAMES[$upgrade_key],
      // ..
  }
}
```

The main benefit I'm thinking this achieves is "if you have an instance of `VillageUpgrade`, you know that the values are there" rather than sometimes having an instance that doesn't have the config values set (from `getUpgradesForVillage`).

Side benefit is isolating some of the complexity out of this method that is more related to how config maps to a `VillageUpgrade` instance, so this class is slightly easier to read through